src/main/java/emu/grasscutter/game/gacha/GachaManager.java

```diff
@@ -296,24 +299,34 @@ public synchronized void doPulls(Player player, int scheduleId, int times) {
 			pools.fallbackItems5Pool1 = removeC6FromPool(pools.fallbackItems5Pool1, player);
 			pools.fallbackItems5Pool2 = removeC6FromPool(pools.fallbackItems5Pool2, player);
 		}
-		
+
 		for (int i = 0; i < times; i++) {
 			// Roll
-			int itemId = doPull(banner, gachaInfo, pools);
-			ItemData itemData = GameData.getItemDataMap().get(itemId);
-			if (itemData == null) {
+            int itemId;
+            ItemData itemData;
+            LootTable table = banner.getLootTable();
+            if (table == null) {
+                itemId = doPull(banner, gachaInfo, pools);
+                itemData = GameData.getItemDataMap().get(itemId);
+            } else {
+                LootContext ctx = playerRecords.computeIfAbsent(player.getUid(), e -> new LootContext());
```

Comment:
This doesn't seem to be serialized, I haven't run this but I expect this to not save banner state between sessions. It also seems to be a single context (i.e. not per banner)?

Reply:
Yes it is not serialized for now. I'll dive into it to find a way to serialize it properly. The general idea of this is instead of saving gacha pull counts in a separated table, we save it as a general data item. The states are meant to saved in a general data store and different banners just use different store keys to know their states.

Reply:
now the data is serialized